services/ingestion/engine.go

```diff
@@ -115,55 +127,69 @@ func (e *Engine) Run(ctx context.Context) error {
 	}
 }
 
-// processEvents iterates all the events and decides based on the type how to process them.
-func (e *Engine) processEvents(events flow.BlockEvents) error {
+// processEvents converts the events to block and transactions and indexes them.
+//
+// BlockEvents are received by the access node API and contain Cadence height (always a single Flow block),
+// and a slice of events. In our case events are EVM events that can contain 0 or multiple EVM blocks and
+// 0 or multiple EVM transactions. The case where we have 0 blocks and transactions is a special heartbeat
+// event that is emitted if there are no new EVM events for a longer period of time
+// (configurable on AN normally a couple of seconds).
+//
+// The values for events payloads are defined in flow-go:
+// https://github.com/onflow/flow-go/blob/master/fvm/evm/types/events.go
+//
+// Any error is unexpected and fatal.
+func (e *Engine) processEvents(blockEvents flow.BlockEvents) error {
 	e.log.Debug().
-		Uint64("cadence-height", events.Height).
-		Int("cadence-event-length", len(events.Events)).
+		Uint64("cadence-height", blockEvents.Height).
+		Int("cadence-event-length", len(blockEvents.Events)).
 		Msg("received new cadence evm events")
 
-	blockEvent := false
-	for _, event := range events.Events {
-		var err error
-		switch {
-		case models.IsBlockExecutedEvent(event.Value):
-			err = e.processBlockEvent(events.Height, event.Value)
-			blockEvent = true
-		case models.IsTransactionExecutedEvent(event.Value):
-			err = e.processTransactionEvent(event.Value)
-		default:
-			return fmt.Errorf("invalid event type") // should never happen
+	events := models.NewCadenceEvents(blockEvents)
+
+	// if heartbeat interval with no data still update the cadence height
+	if events.Empty() {
+		if err := e.blocks.SetLatestCadenceHeight(events.CadenceHeight()); err != nil {
+			return fmt.Errorf("failed to update to latest cadence height during events ingestion: %w", err)
 		}
+		return nil // nothing else to do this was heartbeat event with not event payloads
+	}
 
-		if err != nil {
-			return fmt.Errorf("failed to process event: %w", err)
+	// we first index evm blocks only then transactions if any present
```

Comment:
Technically, unlike Flow, there is no EVM block formation without EVM transactions. Is there a specific reason for indexing in this order? Just curious.

Reply:
this was a bug before, it could be that in the event payload itself we would first get tx and then block event, this just makes sure we first index blocks, because tx references block height and if block is fetched in between by that height it might not yet exist. This is noticeable in even streaming

Reply:
That's a good reason indeed :ok_hand: